libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -699,16 +698,35 @@ string YulUtilFunctions::storageArrayPopFunction(ArrayType const& _type)
 				let oldLen := <fetchLength>(array)
 				if iszero(oldLen) { invalid() }
 				let newLen := sub(oldLen, 1)
-
 				let slot, offset := <indexAccess>(array, newLen)
 				<setToZero>(slot, offset)
-
-				sstore(array, newLen)
+				<?isByteArray>
+					switch oldLen
+					case 32 {
+						// Here we have a special case where array transitions to shorter than 32
+						// So we need to copy data
+						mstore(0, array)
+						let copyFromSlot := keccak256(0, 0x20)
+						let dataToCopy := sload(copyFromSlot)
+						sstore(copyFromSlot, 0)
+						// New length is 31, encoded to 31 * 2 = 62
+						dataToCopy := or(<shl>(dataToCopy), 62)
+						sstore(array, dataToCopy)
+					}
+					default {
+						let data := sload(array)
```

Comment:
We are doing two `sload`s here. Maybe it would be better to have a fully specialized implementation for byte arrays?

Reply:
I made two separate functions now